### PR TITLE
[core] add BitVector template class

### PR DESCRIPTION
**Pull request by simonlingoogle**
This adds the `BitVector` template class and changes `ChildMask` to inherit from `BitVector`.
This can help reduce redundant code for bit mask manipulations.

### CODE REVIEW

File: src/core/thread/child_mask.hpp [hunk 1]
```diff
@@ -36,11 +36,7 @@
 
 #include "openthread-core-config.h"
 
-#include <openthread/error.h>
-
-#include "common/code_utils.hpp"
-#include "common/debug.hpp"
-#include "common/encoding.hpp"
+#include "utils/bit_vector.hpp"
```
Comment (abtink) [inline src/core/thread/child_mask.hpp:39]: I suggest we move `bit_vector.hpp` under `common` folder. The `common` is for all common helper modules used in OT core.

The  `utils` folder is mainly for extra features/functionalities that may not be directly part of Thread spec (and are not directly used by other OT core modules) but are useful additional/extra features we want to include in OT, for example, jam-detector, channel-monitor, heap managment, flash managment, etc.

File: src/core/utils/bit_vector.hpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+/**
+ * This class represents a bit-vector.
+ *
+ */
+template <uint16_t N> class BitVector
+{
+public:
+    /**
+     * This method returns if a given index is masked.
```
Comment (abtink) [inline src/core/utils/bit_vector.hpp:61]:
```suggestion
     * This method indicates whether a given index is included in the mask or not.
```

File: src/core/thread/child_mask.hpp [hunk 3]
```diff
@@ -58,78 +54,8 @@ namespace ot {
  * This class represents a bit-vector of child mask.
  *
  */
-class ChildMask
+class ChildMask : public BitVector<OPENTHREAD_CONFIG_MLE_MAX_CHILDREN>
```
Comment (abtink) [inline src/core/thread/child_mask.hpp:57]: How about using `typedef` here?
```
typedef BitVector<OPENTHREAD_CONFIG_MLE_MAX_CHILDREN> ChildMask;
```

We can make it a sub-class later if/when we have more APIs that we want to add to `ChildMask`?

File: src/core/thread/child_mask.hpp (hunk 1 above)
Comment (simonlingoogle) [reply src/core/thread/child_mask.hpp:39]: I moved it from `common/` to `utils/`. Now I feel sorry for that.

**Comment by simonlingoogle** [comment]
#5236 depends on this PR to be merged to rebase. @librasungirl @jwhui 
Please review when you have time. Thanks

File: src/core/common/bit_vector.hpp [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+/**
+ * @file
+ *   This file includes definitions for a bit-vector.
+ */
+
+#ifndef BIT_VECTOR_HPP_
+#define BIT_VECTOR_HPP_
+
+#include "../openthread-core-config.h"
```
Comment (librasungirl) [inline src/core/common/bit_vector.hpp:37]:
```suggestion
#include "openthread-core-config.h"
```

**Comment by simonlingoogle** [comment]
> LGTM
> 
> As in #5276, ChildMask is required to be comparable and clearable, do you plan to enhance this one or a separate one later?

`BitVector` has a `ClearAll()` method to clear all indexes. 
```
void ClearAll(void) { memset(mMask, 0, sizeof(mMask)); }
```

I intentionally didn't inherit Clearable<BitVector> to avoid using the same method `Clear` for two different methods to reduce confusion. 
Thoughts? @librasungirl 

BTW, Comparable is a good idea, I will add.

File: src/core/common/bit_vector.hpp [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+ * @brief
+ *   This module includes definitions for bit-vector.
+ *
+ * @{
+ *
+ */
+
+/**
+ * This class represents a bit-vector.
+ *
```
Comment (jwhui) [inline src/core/common/bit_vector.hpp:55]: nit: add `@tparam` doc

**Pull request merged**